OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs

```diff
@@ -194,22 +194,105 @@ void IBotTick.BotTick(IBot bot)
 			AssignRolesToIdleUnits(bot);
 		}
 
-		internal Actor FindClosestEnemy(WPos pos)
+		internal static Actor ClosestTo(IEnumerable<Actor> ownActors, Actor targetActor)
 		{
-			var units = World.Actors.Where(IsPreferredEnemyUnit);
-			return units.Where(IsNotHiddenUnit).ClosestTo(pos) ?? units.ClosestTo(pos);
+			// Return actors that can get within weapons range of the target.
+			// First, let's determine the max weapons range for each of the actors.
+			var target = Target.FromActor(targetActor);
+			var ownActorsAndTheirAttackRanges = ownActors
+				.Select(a => (Actor: a, AttackBases: a.TraitsImplementing<AttackBase>().Where(Exts.IsTraitEnabled)
+					.Where(ab => ab.HasAnyValidWeapons(target)).ToList()))
+				.Where(x => x.AttackBases.Count > 0)
+				.Select(x => (x.Actor, Range: x.AttackBases.Max(ab => ab.GetMaximumRangeVersusTarget(target))))
+				.ToDictionary(x => x.Actor, x => x.Range);
+
+			// Now determine if each actor can either path directly to the target,
+			// or if it can path to a nearby location at the edge of its weapon range to the target
+			// A thorough check would check each position within the circle, but for performance
+			// we'll only check 8 positions around the edge of the circle.
+			// We need to account for the weapons range here to account for units such as boats.
+			// They can't path directly to a land target,
+			// but might be able to get close enough to shore to attack the target from range.
+			return ownActorsAndTheirAttackRanges.Keys
+				.ClosestToWithPathToAny(targetActor.World, a =>
+				{
+					var range = ownActorsAndTheirAttackRanges[a].Length;
+					var rangeDiag = Exts.MultiplyBySqrtTwoOverTwo(range);
+					return new[]
+					{
+						targetActor.CenterPosition,
+						targetActor.CenterPosition + new WVec(range, 0, 0),
+						targetActor.CenterPosition + new WVec(-range, 0, 0),
+						targetActor.CenterPosition + new WVec(0, range, 0),
+						targetActor.CenterPosition + new WVec(0, -range, 0),
+						targetActor.CenterPosition + new WVec(rangeDiag, rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(-rangeDiag, rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(-rangeDiag, -rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(rangeDiag, -rangeDiag, 0),
```

Comment:
I'm still unsure how to feel about this change. I feel some sort of backup system would still be needed to account for edge cases

Reply:
Edge cases could be something like a river, more chocky areas, or just obstacles miraculously blocking all the 9 cells without actually blocking the target

Reply:
If an unfortunate terrain layout means a unit can't be targetted, then yes this will (incorrectly) ignore that unit. If it's the only possible option then indeed we would ignore it in this edge case.

However the following mitigate this scenario in my mind:
- If there's other units, the squad can attack those instead.
- If there's another squad with a different attack range, it might not suffer from the same unfortunate terrain setup and be able to attack.
- If there's another squad in the same domain (e.g. they are on the same island as the target) they can attack the target once available.
- If there's aircraft, they can mop up this target.

The required adverse terrain setup is quite specific that it should be very rare in practise, and if it does occur I think any of the above options mean it shouldn't matter really.